File: Spring Week 2/examples/maxflow.py


"""
    *
 * This implementation of Dinic's algorithm is taken from
 * https://python.algorithmexamples.com/web/graphs/dinic.html. We will use
 * it to solve the Kattis problem: https://open.kattis.com/problems/maxflow.
 * Details on how to use this implementation are demonstrated below.
"""
INF = (2 ** 31) - 1
 
 
class Dinic:
    def __init__(self, n):
        self.lvl = [0] * n
        self.ptr = [0] * n
        self.q = [0] * n
        self.adj = [[] for _ in range(n)]
 
    """
    Here we will add our edges containing with the following parameters:
    vertex closest to source, vertex closest to sink and flow capacity
    through that edge ...
    """
 
    def add_edge(self, a, b, c, rcap=0):
        self.adj[a].append([b, len(self.adj[b]), c, 0])
        self.adj[b].append([a, len(self.adj[a]) - 1, rcap, 0])
 
    # This is a sample depth first search to be used at max_flow
    def depth_first_search(self, vertex, sink, flow):
        if vertex == sink or not flow:
            return flow
 
        for i in range(self.ptr[vertex], len(self.adj[vertex])):
            e = self.adj[vertex][i]
            if self.lvl[e[0]] == self.lvl[vertex] + 1:
                p = self.depth_first_search(e[0], sink, min(flow, e[2] - e[3]))
                if p:
                    self.adj[vertex][i][3] += p
                    self.adj[e[0]][e[1]][3] -= p
                    return p
            self.ptr[vertex] = self.ptr[vertex] + 1
        return 0
 
    # Here we calculate the flow that reaches the sink
    def max_flow(self, source, sink):
        flow, self.q[0] = 0, source
        for l in range(31):  # noqa: E741  l = 30 maybe faster for random data
            while True:
                self.lvl, self.ptr = [0] * len(self.q), [0] * len(self.q)
                qi, qe, self.lvl[source] = 0, 1, 1
                while qi < qe and not self.lvl[sink]:
                    v = self.q[qi]
                    qi += 1
                    for e in self.adj[v]:
                        if not self.lvl[e[0]] and (e[2] - e[3]) >> (30 - l):
                            self.q[qe] = e[0]
                            qe += 1
                            self.lvl[e[0]] = self.lvl[v] + 1
 
                p = self.depth_first_search(source, sink, INF)
                while p:
                    flow += p
                    p = self.depth_first_search(source, sink, INF)
 
                if not self.lvl[sink]:
                    break
 
        return flow

    

# End of Dinic's implementation

# Read in the number of nodes and edges
n, m, s, t = map(int, input().split())

# Create a new Dinic object with n nodes
dinic = Dinic(n)

# Read in the edges and add them to the Dinic object
for i in range(m):
    u, v, cap = map(int, input().split())
    # add edge from u to v with capacity 'cap'
    dinic.add_edge(u, v, cap)

# Compute the maximum flow from the source to the sink
maxFlow = dinic.max_flow(s, t)
edges_used = []
# Iterate through all the edges in the flow network
# An edge is used if its flow is greater than 0
# We can also check the capacity of the edge in the same way
for i in range(n):
    for e in dinic.adj[i]:
        if e[3] > 0:
            edges_used.append([i, e[0], e[3]])

print(n, maxFlow, len(edges_used))
for e in edges_used:
    print(e[0], e[1], e[2])

File: Spring Week 2/examples/maxflow.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * This implementation of Dinic's algorithm is taken from
 * https://cp-algorithms.com/graph/dinic.html. We will use
 * it to solve the Kattis problem: https://open.kattis.com/problems/maxflow.
 * Details on how to use this implementation are demonstrated in the main method.
 */ 

//https://cp-algorithms.com/graph/dinic.html

struct FlowEdge {
    int v, u;
    long long cap, flow = 0;
    FlowEdge(int v, int u, long long cap) : v(v), u(u), cap(cap) {}
};

struct Dinic {
    const long long flow_inf = 1e18;
    vector<FlowEdge> edges;
    vector<vector<int>> adj;
    int n, m = 0;
    int s, t;
    vector<int> level, ptr;
    queue<int> q;

    Dinic(int n, int s, int t) : n(n), s(s), t(t) {
        adj.resize(n);
        level.resize(n);
        ptr.resize(n);
    }

    void add_edge(int v, int u, long long cap) {
        edges.emplace_back(v, u, cap);
        edges.emplace_back(u, v, 0);
        adj[v].push_back(m);
        adj[u].push_back(m + 1);
        m += 2;
    }

    bool bfs() {
        while (!q.empty()) {
            int v = q.front();
            q.pop();
            for (int id : adj[v]) {
                if (edges[id].cap - edges[id].flow < 1)
                    continue;
                if (level[edges[id].u] != -1)
                    continue;
                level[edges[id].u] = level[v] + 1;
                q.push(edges[id].u);
            }
        }
        return level[t] != -1;
    }

    long long dfs(int v, long long pushed) {
        if (pushed == 0)
            return 0;
        if (v == t)
            return pushed;
        for (int& cid = ptr[v]; cid < (int)adj[v].size(); cid++) {
            int id = adj[v][cid];
            int u = edges[id].u;
            if (level[v] + 1 != level[u] || edges[id].cap - edges[id].flow < 1)
                continue;
            long long tr = dfs(u, min(pushed, edges[id].cap - edges[id].flow));
            if (tr == 0)
                continue;
            edges[id].flow += tr;
            edges[id ^ 1].flow -= tr;
            return tr;
        }
        return 0;
    }

    long long flow() {
        long long f = 0;
        while (true) {
            fill(level.begin(), level.end(), -1);
            level[s] = 0;
            q.push(s);
            if (!bfs())
                break;
            fill(ptr.begin(), ptr.end(), 0);
            while (long long pushed = dfs(s, flow_inf)) {
                f += pushed;
            }
        }
        return f;
    }
};
/*
* End of Dinic's algorithm implementation
*/

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    // Read in the number of nodes and edges
    int n, m, s, t;
    cin >> n >> m >> s >> t;

    // Create a new Dinic object with n nodes, source node s, and sink node t
    Dinic dinic(n, s, t);

    // Read in the edges and add them to the Dinic object
    for (int i = 0; i < m; i++) {
        int u, v, cap;
        cin >> u >> v >> cap;
        // add edge from u to v with capacity 'cap'
        dinic.add_edge(u, v, cap);
    }

    // Compute the maximum flow from the source to the sink
    long long maxFlow = dinic.flow();
    vector<FlowEdge> edges_used;
    // Iterate through all the edges in the flow network
    // An edge is used if its flow is greater than 0
    // We can also check the capacity of the edge in the same way
    for (FlowEdge e : dinic.edges) {
        if (e.flow > 0) {
            edges_used.push_back(e);
        }
    }
    cout << n << ' ' << maxFlow << ' ' << edges_used.size() << '\n';
    for (FlowEdge e : edges_used) {
        cout << e.v << ' ' << e.u << ' ' << e.flow << '\n';
    }
}
 